### PR TITLE
Adding revalidation of StoreFindToken

### PR DESCRIPTION
**Pull request by vgkholla**
StoreFindTokens may no longer be valid after a compaction (because the index segment or offset it refers to no longer exists). This commit ensures that such tokens are reset.

Reviewers: @pnarayanan, @nsivabalan

All new lines  (except the `default` case in the `switch` in `revalidateStoreFindToken()`) are covered.

### CODE REVIEW

File: ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java [hunk 1]
```diff
@@ -1319,6 +1326,52 @@ void persistIndex() throws StoreException {
   }
 
   /**
+   * Re-validates the {@code token} to ensure that it is consistent with the current set of index segments. If it is
+   * not, a {@link StoreFindToken.Type#Uninitialized} token is returned.
+   * @param token the {@link StoreFindToken} to revalidate.
+   * @return {@code token} if is consistent with the current set of index segments, a new
+   * {@link StoreFindToken.Type#Uninitialized} token otherwise.
+   */
+  FindToken revalidateFindToken(FindToken token) {
```
Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java]: do we need to check for revalidation for getBlobReadInfo(). I mean, once we get the indexValue, is there a chance that the log entry was compacted?

Comment (nsivabalan) [inline ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java]: Also, don't we need to revalidate in findDeletedEntriesSince() apart from findEntriesSince()

File: ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java [hunk 2]
```diff
@@ -173,6 +173,8 @@ void resume() {
     hardDeleteLock.lock();
     try {
       if (paused.compareAndSet(true, false)) {
+        startToken =
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java]: I think this is unrelated to this change and the check you suggest should be added as a test assert in the hard delete pause-resume test in `IndexTest`. When you find the time, could you add the assertion?

File: ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java [hunk 3]
```diff
@@ -699,6 +699,8 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
       Offset logEndOffsetBeforeFind = log.getEndOffset();
       StoreFindToken storeToken = resetTokenIfRequired((StoreFindToken) token);
       logger.trace("Time used to validate token: {}", (time.milliseconds() - startTimeInMs));
+      ConcurrentSkipListMap<Offset, IndexSegment> indexSegments = validIndexSegments;
+      storeToken = revalidateStoreFindToken(storeToken, indexSegments);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java]: `resetTokenIfRequired()` is done only once at the beginning of the function. Revalidating store find token however is done whenever a new view is obtained. So they should be distinct functions.

File: ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java (hunk 1 above)
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java]: There is no revalidation required in `getBlobReadInfo()` (also it has no tokens).

`findDeletedEntriesSince()` seems to have been written with the assumption that it is used only internally in the store and its only use case is hard delete. It does not call `resetTokenIfRequired()` either. If we want to bring it up to par with `findEntriesSince()` (there is no reason to do that currently), we should do all the required things (out of scope for this change). Adding only the revalidate is not of much use.

File: ambry-store/src/test/java/com.github.ambry.store/IndexTest.java [hunk 4]
```diff
@@ -798,6 +798,37 @@ public void getLogSegmentsNotInJournalTest() throws InterruptedException, IOExce
   }
 
   /**
+   * Tests the cases where tokens have to be revalidated on a call to
+   * {@link PersistentIndex#findEntriesSince(FindToken, long)}.
+   * @throws StoreException
+   */
+  @Test
+  public void tokenRevalidationTest() throws StoreException {
```
Comment (nsivabalan) [inline ambry-store/src/test/java/com.github.ambry.store/IndexTest.java]: Once this is merged, can you add a test to BlobCompactorTest or IndexTest as to check that hard delete tokens are reset if compactor has compated the entry which the hard delete token was referring to. Also other case too, if not for compaction, hard delete should proceed as usual even if there are segments which was compacted as future segments.

Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/IndexTest.java]: That sounds like an integration test - will give it a shot and see if it can be done deterministically (since hard delete is a thread of its own).

**Comment by vgkholla** [comment]
@nsivabalan Thanks for the quick review - if everything looks good to you, you could approve the changes.

**Comment by nsivabalan** [comment]
Looks good to me.
